mdcopy/mdcopy.go

```diff
@@ -0,0 +1,36 @@
+package mdcopy
+
+import (
+	"context"
+	"fmt"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// UnaryServerInterceptor returns a server interceptor
+// that copies request metadata into response metadata.
+func UnaryServerInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+		// Copy the metadata
+		if err := copyMetadata(ctx); err != nil {
+			fmt.Println("Failed to copy metadata:", err)
+		}
+		// Proceed with the handler
+		return handler(ctx, req)
+	}
+}
+
+func copyMetadata(ctx context.Context) error {
+	// Extract incoming metadata from the context
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		fmt.Println("No incoming metadata found")
+		return nil
+	}
+	// Set the response header metadata
+	if err := grpc.SetHeader(ctx, md); err != nil {
```

Comment:
I think we should filter this like the http WithOutgoingHeaderMatcher(). Otherwise, all the pairs in the meta will be copied to the response.

Reply:
My initial thinking was that the outgoing header matcher will filter out the response headers anyways, but since we want to be able to filter the metadata itself I changed the interceptor to accept a map.